File: 2021/csharp/Aoc2021/Day09.cs
namespace Aoc2021;

public class Day09 : IDay
{
    public override string Day { get; } = nameof(Day09)[3..];

    public override string Result1()
    {
        return RiskLevelSum(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return BasinProduct(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    private static int[][] BuildGrid(IEnumerable<string> lines)
    {
        return lines
            .Select(line => line.ToCharArray().Select(c => c - '0').ToArray())
            .ToArray();
    }

    private static IEnumerable<(int x, int y, int height)> LowPoints(int[][] grid)
    {
        int yLength = grid.Length;
        int xLength = grid[0].Length;

        for (int y = 0; y < yLength; y++)
        {
            for (int x = 0; x < xLength; x++)
            {
                int height = grid[y][x];
                if (AdjacentPoints(x, y, grid)
                    .Select(p => grid[p.y][p.x])
                    .All(h => height < h))
                {
                    yield return (x, y, height);
                }
            }
        }
    }

    private static IEnumerable<(int x, int y)> AdjacentPoints(int x, int y, int[][] grid)
    {
        int yLength = grid.Length;
        int xLength = grid[0].Length;

        return new (int x, int y)[]
        {
            (x-1, y),
            (x+1, y),
            (x, y-1),
            (x, y+1),
        }
        .Where(p => p.x >= 0 && p.x < xLength && p.y >= 0 && p.y < yLength);

    }

    public static int RiskLevelSum(IEnumerable<string> lines)
    {
        int[][] grid = BuildGrid(lines);
        return LowPoints(grid).Sum(item => item.height + 1);
    }

    public static int BasinProduct(IEnumerable<string> lines)
    {
        int[][] grid = BuildGrid(lines);

        return LowPoints(grid)
            .Select(p => basinSize(p.x, p.y))
            .OrderByDescending(size => size)
            .Take(3)
            .Aggregate(1, (acc, v) => acc * v);

        int basinSize(int x, int y)
        {
            HashSet<(int, int)> basinPoints = new() { (x, y) };
            buildBasin(x, y, basinPoints);
            return basinPoints.Count;
        }

        void buildBasin(int x, int y, HashSet<(int, int)> basinPoints)
        {
            int height = grid[y][x];
            foreach ((int x, int y) p in AdjacentPoints(x, y, grid))
            {
                if (!basinPoints.Contains(p) && grid[p.y][p.x] != 9)
                {
                    _ = basinPoints.Add(p);
                    buildBasin(p.x, p.y, basinPoints);
                }
            }
        }
    }
}

File: 2021/fsharp/Aoc2021/Day09.fs
module Day09

open Utils

type Point = int * int
type Grid = int array array

module Grid =
    let build (lines: string list) =
        lines
        |> List.map (fun line -> line.ToCharArray() |> Array.map (fun c -> int c - int '0'))
        |> List.toArray

    let height (grid: Grid) (x, y) = grid.[y].[x]

    let dimensions (grid: Grid) = grid.[0].Length, grid.Length

    let contains (grid: Grid) (x, y) =
        let nx, ny = grid |> dimensions
        x >= 0 && x < nx && y >= 0 && y < ny

    let adjacentPoints (grid: Grid) (x, y) =
        [ x - 1, y; x + 1, y; x, y - 1; x, y + 1 ] |> List.filter (grid |> contains)

    let lowPoints (grid: Grid) =
        let nx, ny = grid |> dimensions

        seq {
            for y = 0 to ny - 1 do
                for x = 0 to nx - 1 do
                    let h = (x, y) |> height grid

                    if (x, y) |> adjacentPoints grid |> List.forall (fun p -> h < (p |> height grid)) then
                        yield x, y
        }

    let basinSize (grid: Grid) p =
        let rec build basinPoints p =
            p
            |> adjacentPoints grid
            |> List.filter (fun p -> basinPoints |> Set.contains p |> not && (p |> height grid) <> 9)
            |> List.fold (fun acc p -> p |> build (acc |> Set.add p)) basinPoints

        p |> build (Set.empty |> Set.add p)

let riskLevelSum lines =
    let grid = lines |> Grid.build

    grid |> Grid.lowPoints |> Seq.sumBy (fun p -> 1 + (p |> Grid.height grid))

let basinProduct lines =
    let grid = lines |> Grid.build

    grid
    |> Grid.lowPoints
    |> Seq.map (Grid.basinSize grid)
    |> Seq.map Set.count
    |> Seq.sortDescending
    |> Seq.take 3
    |> Seq.fold (fun acc value -> acc * value) 1

let input = "09" |> readInputLines |> Seq.toList

let job1 () = input |> riskLevelSum |> string

let job2 () = input |> basinProduct |> string
